Support "massive" elliptic DG operator

nilsvu opened the pull request:
## Proposed changes

Add the option to multiply the elliptic DG operator by the mass matrix. The "massive" operator can be easier to solve because it is symmetric, or at least closer to symmetry.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Support massive elliptic DG operator

Optionally apply the DG mass matrix in diagonal approximation
to the elliptic DG operator. The massive DG operator can be
easier to solve because it is symmetric, or at least closer
to symmetry.

nilsvu committed:
Add projection for massive quantities

The L2 projection for massive quantities is just a transpose interpolation.

nilsvu committed:
Add DG apply_mass_matrix function

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Mass.hpp:
```diff
@@ -0,0 +1,76 @@
+ * \f}
+ *
+ * where \f$d\f$ is the spatial dimension and \f$w_{p_i}\f$ are the quadrature
+ * weights in dimension \f$i\f$. To apply the mass in coordinates different than
```
`apply the mass` -> `apply the mass matrix`?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Mass.hpp:
```diff
@@ -0,0 +1,76 @@
+ * logical, or to account for a curved background metric, the data can be
+ * pre-multiplied with the Jacobi determinant and/or a metric determinant.
+ *
+ * Note that the mass-lumping approximation is exact on Legendre-Gauss meshes,
```
omit `approximation` since it's not approximate if it's exact?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Mass.hpp:
```diff
@@ -0,0 +1,76 @@
+ */
+// @{
+template <size_t Dim>
+void apply_mass(const gsl::not_null<DataVector*> data,
```
`apply_mass_matrix` (and below)

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Mass.hpp:
```diff
@@ -0,0 +1,76 @@
+
+namespace detail {
+template <size_t Dim>
+void apply_mass_impl(const gsl::not_null<double*> data,
```
`apply_mass_matrix_impl`, remove `const` in decl

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_Mass.cpp:
```diff
@@ -0,0 +1,108 @@
+namespace {
+
+template <size_t Dim, typename... CoordMaps>
+void test_mass(const Mesh<Dim>& mesh,
```
`test_mass` -> `test_mass_matrix`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_Mass.cpp:
```diff
@@ -0,0 +1,108 @@
+               const domain::CoordinateMap<Frame::Logical, Frame::Inertial,
+                                           CoordMaps...>& coordinate_map,
+               const DataVector& scalar_field,
+               const DataVector& expected_massive_scalar_field) noexcept {
```
a `massive_scalar_field` is a thing that is not this thing as far as my understanding goes. How about `expected_mass_matrix_times_scalar_field`?

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_Mass.cpp:
```diff
@@ -0,0 +1,108 @@
+    const auto coord_map =
+        domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(
+            Affine{-1.0, 1.0, -0.3, 0.7});
+    test_mass(mesh, coord_map, {1., 2., 3., 4.},
```
is there a way to compute the mass matrix from the nodal to modal matrices or something? having a bunch of magic numbers in a test file isn't fabulous :shrug:

nilsvu replied:
Added a test with the mass matrix computed from the Vandermonde matrix. Note that it's only equivalent for Gauss quadrature, because of the approximation made for Gauss-Lobatto quadrature. Therefore I kept the hard-coded numbers around for Gauss-Lobatto quadrature. They are just the data divided by the quadrature weights, so I added a comment.

nilsdeppe commented on src/Elliptic/Actions/InitializeFixedSources.hpp:
```diff
@@ -76,6 +79,15 @@ struct InitializeFixedSources {
     auto fixed_sources = variables_from_tagged_tuple(background.variables(
         inertial_coords, typename fixed_sources_tag::type::tags_list{}));
 
+    // Apply DG mass to the fixed sources if the DG operator is massive
```
`mass` -> `mass matrix`

nilsdeppe commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -850,6 +853,12 @@ struct DgOperatorImpl<System, Linearized, tmpl::list<PrimalFields...>,
                         mesh.extents(), direction.dimension(), slice_index);
     }  // loop over all mortars
+
+    // Apply DG mass
```
`mass` -> `mass matrix`

nilsdeppe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:
```diff
@@ -150,7 +151,7 @@ template <
 void test_dg_operator(
     const DomainCreator<Dim>& domain_creator, const double penalty_parameter,
-    const AnalyticSolution& analytic_solution,
+    const bool massive, const AnalyticSolution& analytic_solution,
```
`massive` -> `use_massive_dg_operator`

nilsdeppe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:
```diff
@@ -613,9 +631,12 @@ SPECTRE_TEST_CASE("Unit.Elliptic.DG.Operator", "[Unit][Elliptic]") {
-      test_dg_operator<system, true>(
-          domain_creator, penalty_parameter, analytic_solution,
-          analytic_solution_aux_approx, analytic_solution_operator_approx, {});
+      for (const bool massive : {true, false}) {
```
`massive` -> `use_massive_dg_operator`

nilsdeppe commented on tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:
```diff
@@ -733,9 +754,12 @@ SPECTRE_TEST_CASE("Unit.Elliptic.DG.Operator", "[Unit][Elliptic]") {
-      test_dg_operator<system, true>(
-          domain_creator, penalty_parameter, analytic_solution,
-          analytic_solution_aux_approx, analytic_solution_operator_approx, {});
+      for (const bool massive : {true, false}) {
```
`massive` -> `use_massive_dg_operator`

nilsdeppe commented on src/NumericalAlgorithms/Spectral/Projection.cpp:
```diff
@@ -49,9 +49,10 @@ bool needs_projection(const Mesh<Dim>& mesh1, const Mesh<Dim>& mesh2,
+const Matrix& projection_matrix_child_to_parent(const Mesh<1>& child_mesh,
+                                                const Mesh<1>& parent_mesh,
+                                                const ChildSize size,
+                                                const bool massive) noexcept {
```
better name please...

Why not make this a separate function? It seems like the only thing shared are `ASSERT`s.

Also, since it's just the transpose, would it make more sense to handle this at the sight where the matrix is applied passing `'T'` to the `dgemm` call? I don't know since I have not used the code

nilsvu replied:
renamed in a fixup. I considered making this a separate function, but then the call site is more complicated. It's nice to be able to use this as `apply_matrices(projection_matrix(mesh, parent_mesh, operand_is_massive), operand, mesh)`.

nilsdeppe replied:
Sounds good :)

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Projection.cpp:
```diff
@@ -342,6 +344,100 @@ void test_h_element_to_mortar() {
+void test_massive_restriction() {
+  INFO("Massive restriction operator");
+  // Using Gauss quadrature so the diagonal mass-matrix approximation used in
+  // `::dg::apply_mass` is exact. Note that for Gauss-Lobatto quadrature the
```
`apply_mass` -> `apply_mass_matrix`

nilsdeppe commented on src/NumericalAlgorithms/Spectral/Projection.hpp:
```diff
@@ -53,6 +53,18 @@ bool needs_projection(const Mesh<Dim>& mesh1, const Mesh<Dim>& mesh2,
+ * transpose of the prolongation operator, i.e. just an interpolation matrix
+ * transpose. Note that the "massive" residual already takes the difference
+ * in element size between parent and children into account by including a
+ * Jacobian in the volume element of the integral.
```
are you certain this holds for curved meshes? I think all our projection currently isn't conservative on curved meshes (when `u J` doesn't fit in the basis to begin with), but I don't know how that translates to "massive" quantities

nilsvu replied:
Projecting a massive quantity essentially means we project `u J`, see e.g. `InitializeFixedSources.hpp` or `DgOperator.hpp` where I multiply by the Jacobian. Not sure if that's conservative, but it's better than what we currently do when projecting between faces and mortars where (I think) we do a "logical" projection, i.e. only have a factor of 2 from the Jacobian between parent and children logical coords. That's correct on a rectilinear grid, but I think for better projections on a curved mesh we should remove the factor of 2 and multiply by Jacobians. Of course that means we have to evaluate and store the Jacobians on the mortars.

nilsdeppe replied:
ah okay, yea if you project `u J` then it's conservative, or, you conserve the inertial frame quantity, which is generally what we want :)

nilsvu commented:
Pushed fixups, and renamed a lot of things (squashed in the renames directly). Thanks for reviewing @nilsdeppe!

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/ApplyMassMatrix.hpp:
```diff
@@ -0,0 +1,77 @@
+ * where \f$d\f$ is the spatial dimension and \f$w_{p_i}\f$ are the quadrature
+ * weights in dimension \f$i\f$. To apply the mass matrix in coordinates
+ * different than logical, or to account for a curved background metric, the
+ * data can be pre-multiplied with the Jacobi determinant and/or a metric
```
`Jacobi` -> `Jacobian`?

nilsdeppe reviewed (commented):
LGTM, just one small typo (I think), and could you rename `apply_mass` to `apply_mass_matrix` in the name of the first commit while squashing? Please go ahead and squash :)

moxcodes commented on src/NumericalAlgorithms/DiscontinuousGalerkin/ApplyMassMatrix.hpp:
```diff
@@ -0,0 +1,77 @@
+ * data can be pre-multiplied with the Jacobian determinant and/or a metric
+ * determinant.
+ *
+ * Note that the mass-lumping is exact on Legendre-Gauss meshes, but omits a
```
Note could be \note

moxcodes commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_ApplyMassMatrix.cpp:
```diff
@@ -0,0 +1,181 @@
+}
+
+template <size_t Dim>
+std::array<Matrix, Dim> exact_logical_mass_matrix(const Mesh<Dim>& mesh) {
```
[nitpick] the other functions are `noexcept` but this one isn't

moxcodes commented on src/Elliptic/DiscontinuousGalerkin/Tags.hpp:
```diff
@@ -51,5 +60,15 @@ struct PenaltyParameter : db::SimpleTag {
+/// Whether or not to multiply the DG operator with the mass matrix. Massive DG
+/// operators can be easier to solve because they are symmetric, or at least
+/// closer to symmetry.
+struct Massive : db::SimpleTag {
```
Analogous to the variable name changes suggested by @nilsdeppe , this tag could be `UseMassiveDgOperator` or `DgOperatorIsMassive` or similar to be more clear

nilsvu replied:
This is listed in the input file like this:

```yaml
Discretization:
  DiscontinuousGalerkin:
    Massive: True
```

so it has the appropriate "DG" context. Same for your comment below: It's in the `dg::apply_operator` function, which provides the context.

moxcodes commented on src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:
```diff
@@ -620,13 +621,15 @@ struct DgOperatorImpl<System, Linearized, tmpl::list<PrimalFields...>,
       const ::dg::MortarMap<Dim, Mesh<Dim - 1>>& all_mortar_meshes,
       const ::dg::MortarMap<Dim, ::dg::MortarSize<Dim - 1>>& all_mortar_sizes,
-      const double penalty_parameter, const TemporalId& temporal_id,
+      const double penalty_parameter, const bool massive,
```
consider making the bool names more clear here as you did in the test.

moxcodes reviewed (dismissed):
The code looks good -- I only have tiny nitpicks that you can squash right away.
I feel, though, that there is a key mathematical realization that I'm missing in this -- I can understand how constructing the linear equation such that the operator is closer to symmetric can make it faster to solve. But, how is it managing to solve the same problem? I would have naively thought that for a given differential equation, it should have a unique linearization. So, why can we apply the mass matrix and end up with a solution to the same problem? Or is there a correction that is applied at the end that I haven't properly grasped?

nilsvu commented:
> I feel, though, that there is a key mathematical realization that I'm missing in this -- I can understand how constructing the linear equation such that the operator is closer to symmetric can make it faster to solve. But, how is it managing to solve the same problem? I would have naively thought that for a given differential equation, it should have a unique linearization. So, why can we apply the mass matrix and end up with a solution to the same problem? Or is there a correction that is applied at the end that I haven't properly grasped?

First, the issue of mass matrices is unrelated to linearizations. We DG-discretize both the nonlinear and the linearized equations so we end up with a nonlinear operator `A(u)` and its linearization `dA/du`. For the purpose of this PR let's assume that `A(u)` is already linear, like for the Poisson or elasticity systems.

Solving the elliptic problem means solving `A(u) = b` for `u`. The solution doesn't change when we remove the DG-mass from both sides: `M^-1 * A(u) = M^-1 * b`. Now `A` can be symmetric but `M^-1 * A` is generally not. Therefore, both converge to the same solution but `A` has better convergence properties than `M^-1 * A`.

Note that I'm writing `A` and `M^-1 * A` because by default `A` contains the mass matrix: we DG-discretize by taking inner products with the basis functions: `\int \psi_p(x) f(x) dx ~= M_pq f_q`. In the mass-lumping approximation it's easy to remove the mass matrix by a simple pointwise division so in the evolution code we generally work with `M^-1 * A` (see e.g. https://arxiv.org/abs/1510.01190 Eq. (3.17)) but in the elliptic code it can be advantageous to keep the mass matrix in the operator (faster convergence due to symmetry). Not sure if such symmetry arguments can also help in the evolution code.

nilsvu commented:
Rebased and squashed in the two minor suggestions. Thanks for reviewing @nilsdeppe and @moxcodes, and I hope I have answered your question @moxcodes :)

nilsvu commented:
Oops accidentally deleted the remote branch, restored now

moxcodes reviewed (approved):
Ah, I see -- I had overlooked that the same mass matrix is being applied to the source as well, thanks for the explanation!
